numpy_minmax/_minmax.c

```diff
@@ -63,22 +60,118 @@ MinMaxResult minmax_avx2(float *a, size_t length) {
     }
 
-    // Reduce min and max values from AVX registers
-    float temp_min[8], temp_max[8];
-    _mm256_storeu_ps(temp_min, min_vals);
-    _mm256_storeu_ps(temp_max, max_vals);
-    for (i = 0; i < 8; ++i) {
-        if (temp_min[i] < result.min_val) result.min_val = temp_min[i];
-        if (temp_max[i] > result.max_val) result.max_val = temp_max[i];
-    }
+    REDUCE_RESULT_FROM_MM256(min_vals, max_vals, result);
 
     return result;
 }
 
 MinMaxResult minmax_contiguous(float *a, size_t length) {
+    // Return early for empty arrays
+    if (length == 0) {
+        return (MinMaxResult){0.0, 0.0};
+    }
     if (length >= 16) {
-        return minmax_avx2(a, length);
+        return minmax_avx(a, length);
     } else {
         return minmax_pairwise(a, length);
     }
 }
+
+// Takes the pairwise min/max on strided input. Strides are in number of bytes,
+// which is why the data pointer is char
+MinMaxResult minmax_pairwise_strided(char *a, size_t length, long stride) {
+    MinMaxResult result;
+
+    // Initialize min and max with the last element of the array.
+    // This ensures that it works correctly for odd length arrays as well as even.
+    result.min_val = *(float*)(a + (length -1)*stride);
+    result.max_val = result.min_val;
+
+    // Process elements in pairs
+    float smaller;
+    float larger;
+    for (size_t i = 0; i < (length - 1)*stride; i += 2*stride) {
+        if (*(float*)(a + i) < *(float*)(a + i + stride)) {
+            smaller = *(float*)(a + i);
+            larger = *(float*)(a + i + stride);
+        } else {
+            smaller = *(float*)(a + i + stride);
+            larger = *(float*)(a + i);
+        }
+
+        if (smaller < result.min_val) {
+            result.min_val = smaller;
+        }
+        if (larger > result.max_val) {
+            result.max_val = larger;
+        }
+    }
+    return result;
+}
+
+// Takes the avx min/max on strided input. Strides are in number of bytes,
+// which is why the data pointer is char
+MinMaxResult minmax_avx_strided(char *a, size_t length, long stride) {
+    MinMaxResult result = { .min_val = FLT_MAX, .max_val = -FLT_MAX };
+
+    __m256 min_vals = _mm256_set_ps(
+        *(float*)(a),
+        *(float*)(a + stride),
+        *(float*)(a + 2*stride),
+        *(float*)(a + 3*stride),
+        *(float*)(a + 4*stride),
+        *(float*)(a + 5*stride),
+        *(float*)(a + 6*stride),
+        *(float*)(a + 7*stride)
+        );
+    __m256 max_vals = min_vals;
+
+    // Process elements in chunks of eight
+    size_t i = 8*stride;
+    for (; i <= (length - 8)*stride; i += 8*stride) {
+        __m256 vals = _mm256_set_ps(
+            *(float*)(a + i),
+            *(float*)(a + i + stride),
+            *(float*)(a + i + 2*stride),
+            *(float*)(a + i + 3*stride),
+            *(float*)(a + i + 4*stride),
+            *(float*)(a + i + 5*stride),
+            *(float*)(a + i + 6*stride),
+            *(float*)(a + i + 7*stride)
```

Comment:
Is there an AVX instruction that can do these multiply-accumulate operations?

Reply:
There is an AVX instruction for multiply-accumulate, but that is not what we are doing here, did you mean something else?

Reply:
Yeah, sorry, I misread the code

Reply:
Or did you mean the indices? Do you think that is worth it?

Reply:
Tested with `_mm256_i32gather_ps` (vgatherdps), and it was slower on my computer.